### ISSUE TITLE
Match-time code evaluation problem with arrays

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7500]
Migrated from [rt.perl.org#31565](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31565) (status was 'resolved')

Searchable as RT31565$

**Comment by p5pRT** [comment]
### From djv@mbnet.fi

#### Created by djv@mbnet.fi



There's basically some strange behaviour concerning visibility
of arrays during match\-time code evaluation\, that I can not fix\.
That's about as well as I can put it\.

Here are test cases depicting the problem\.
The script wants input lines of style "\\d\*\(\(\, \\d\*\) and \\d\*\)?"
it then prints all the "\\d\*"s\.
The latter test case fails on all lines after the first one \-
there are no items in @&#8203;list\.
\-\-\- works always \-\->
\#\!/usr/bin/perl \-w

my @&#8203;list=\(\);

while\( my $line=\<STDIN>\) \{
&nbsp;   $line =~ s/\(\\r|\\n\)\*$//;
&nbsp;   $line =~ s/^\(\\d\*\)\(?&#8203;:\(?&#8203;:\, \(\\d\*\)\(?\{push @&#8203;list\, $2 if defined $2;\}\)\)\* and 
\(\\d\*\)\)?$/\\n/;
&nbsp;   if\( $line eq "\\n"\) \{
&nbsp;     print "\\$1&#8203;: \\"$1\\"\\n";
&nbsp;     my $three=$3;
&nbsp;     foreach my $item \(@&#8203;list\) \{
&nbsp;       print "\\$2&#8203;: \\"$item\\"\\n";
&nbsp;     \}
&nbsp;     print "\\$3&#8203;: \\"$three\\"\\n"    if defined\( $three\);
&nbsp;   \}
&nbsp;   @&#8203;list=\(\);
\}
\<\-\-
fails on second and subsequent lines \-\->
\#\!/usr/bin/perl \-w

while\( my $line=\<STDIN>\) \{
&nbsp;   my @&#8203;list=\(\);
&nbsp;   $line =~ s/\(\\r|\\n\)\*$//;
&nbsp;   $line =~ s/^\(\\d\*\)\(?&#8203;:\(?&#8203;:\, \(\\d\*\)\(?\{push @&#8203;list\, $2 if defined $2;\}\)\)\* and 
\(\\d\*\)\)?$/\\n/;
&nbsp;   if\( $line eq "\\n"\) \{
&nbsp;     print "\\$1&#8203;: \\"$1\\"\\n";
&nbsp;     my $three=$3;
&nbsp;     foreach my $item \(@&#8203;list\) \{
&nbsp;       print "\\$2&#8203;: \\"$item\\"\\n";
&nbsp;     \}
&nbsp;     print "\\$3&#8203;: \\"$three\\"\\n"    if defined\( $three\);
&nbsp;   \}
&nbsp;   @&#8203;list=\(\);
\}
\<\-\-\-

perl v5\.6\.0 built for sun4\-solaris also affected

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=medium

Site configuration information for perl v5.8.4:

Configured by ActiveState at Tue Jun  1 11:52:09 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
   Platform:
     osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
     uname=''
     config_args='undef'
     hint=recommended, useposix=true, d_sigaction=undef
     usethreads=undef use5005threads=undef useithreads=define usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE 
-DNO_STRICT -DHAVE_DES_FCRYPT  -DNO_HASH_SEED -DPERL_IMPLICIT_CONTEXT 
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
     optimize='-MD -Zi -DNDEBUG -O1',
     cppflags='-DWIN32'
     ccversion='', gccversion='', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf 
-libpath:"C:\Perl\lib\CORE"  -machine:x86'
     libpth=C:\PROGRA~1\MICROS~3\VC98\lib
     libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib 
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib 
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib 
msvcrt.lib
     perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib 
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib 
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib 
msvcrt.lib
     libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
     gnulibc_version='undef'
   Dynamic Linking:
     dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf 
-libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
     ACTIVEPERL_LOCAL_PATCHES_ENTRY
     22751 Update to Test.pm 1.25
     21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.4:
     C:/Perl/lib
     C:/Perl/site/lib
     .


Environment for perl v5.8.4:
     HOME (unset)
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
 
PATH=C:\Tcl\bin;C:\Perl\bin\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;c:\Program 
Files\vim\runtime;C:\Program 
Files\Java\j2re1.4.2_04\bin;d:\pak;d:\pak\infozip;d:\pak\gnu;d:\pak\GnuPG;d:\pak\PKZip;d:\pak\jikes;d:\pak\arj;C:\Program 
Files\VDMSound\;C:\Program Files\SSH Communications Security\SSH Secure 
Shell;C:\Program Files\XVI32;C:\Program Files\WinPcap;C:\Program Files\Common 
Files\GTK\2.0\bin;C:\Program Files\CVSNT;d:\batit;c:\Program Files\D-Tools
     PERL_BADLANG (unset)
     SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Sep 14\, 2004 at 04&#8203;:58&#8203;:19PM \-0000\, Tuomo Latto wrote&#8203;:
> There's basically some strange behaviour concerning visibility
> of arrays during match\-time code evaluation\, that I can not fix\.
> That's about as well as I can put it\.

Thanks for the report\. There are well\-known problems associated with
using lexical variables within \(?\{\.\.\.\}\)'s\. Currently the best workaround
is to use global variables within them instead\.

The problem you are seeing is that at compile\-time\, the @&#8203;list within
the regex is bound to the first instance of the @&#8203;list variable\. Each time
round the loop\, a new instance is created at the 'my @&#8203;list'
statement\, but the regex code still sees the old copy of the variable\.

Eg the following code

&nbsp;    my @&#8203;r;
&nbsp;    while \(\<DATA>\) \{
&nbsp;       my @&#8203;a;
&nbsp;       push @&#8203;r\, \\@&#8203;a;
&nbsp;       /\(\\w\+\)\(?\{push @&#8203;a\, $1\}\)/;
&nbsp;    \}
&nbsp;    print "\(@&#8203;$\_\)\\n" for @&#8203;r;
&nbsp;    \_\_DATA\_\_
&nbsp;    one
&nbsp;    two
&nbsp;    three

outputs

&nbsp;    \(one two three\)
&nbsp;    \(\)
&nbsp;    \(\)

As you can see\, the regex keeps pushing the values onto the first instance
of @&#8203;list\.

I'm hoping to change this behaviour in Perl 5\.10\.0\, if I get the time\.

Dave\.


\-\- 
But Pity stayed his hand\. "It's a pity I've run out of bullets"\,
he thought\. \- "Bored of the Rings"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
